### ISSUE TITLE
Gradle 5.x + gradle-versions-plugin/dependencyUpdates = StackOverflowError

### ISSUE THREAD

**Issue by danbar74** [gradle/gradle#9119]
Running `dependencyUpdates` (https://github.com/ben-manes/gradle-versions-plugin) with Gradle 5.x (tried a few different versions) results with StackOverflowError. Originally, I've submitted a bug for gradle-versions-plugin, but it seems it is actually a problem with Gradle (see https://github.com/ben-manes/gradle-versions-plugin/issues/306 for more details).

### Expected Behavior
The `dependencyUpdates` task should work fine with Gradle 5.x (as it does with Gradle 4.x).

### Current Behavior
The `dependencyUpdates` results with StackOverflowError. Note, that regular build tasks (clean/check/build) work fine.

### Steps to Reproduce 
Cannot provide a self-contained project demoing the issue. It seems to be related to the fact that the project uses a private repository/cache (Artifactory) with manually deployed snapshot build of Liquibase (note it is only a problem for Gradle 5.x).

The build file (it's from production code, so I had to remove a few post-build tasks):
[build.txt](https://github.com/gradle/gradle/files/3082258/build.txt)

Gradle output when running `dependencyUpdates` (with Liquibase version set to 3.6.2):
[dependencyUpdates-success-gradle-4.10.3.txt.zip](https://github.com/gradle/gradle/files/3082313/dependencyUpdates-success-gradle-4.10.3.txt.zip)
[dependencyUpdates-stackoverflowerror-gradle-5.3.1.txt.zip](https://github.com/gradle/gradle/files/3082315/dependencyUpdates-stackoverflowerror-gradle-5.3.1.txt.zip)

Gradle output (5.x) when Liquibase version is set to '+' (mentioning this, cause the behaviour is different). Note the stacktrace from refreshing the project in IntelliJ IDEA:
[dependencyUpdates-gradle-5.3.1.txt.zip](https://github.com/gradle/gradle/files/3082275/dependencyUpdates-gradle-5.3.1.txt.zip)
[refresh-project-in-intellij-idea-gradle-53.1.txt.zip](https://github.com/gradle/gradle/files/3082276/refresh-project-in-intellij-idea-gradle-53.1.txt.zip)

### Your Environment
macOS 10.14.4
JDK 1.8.0_192

**Comment by danbar74** [comment]
Tested with Gradle 5.4 - still the same issue.

**Comment by big-guy** [comment]
Can you run with `--warning-mode=all` and paste what the deprecation warnings were?

**Comment by danbar74** [comment]
```
./gradlew dependencyUpdates --warning-mode all

> Configure project :backend
The append property has been deprecated. This is scheduled to be removed in Gradle 6.0. Append should always be true.
        at build_5hzmut2s6xz1df4vudadanqqz$_run_closure8$_closure43.doCall([...]/build.gradle:195)
        (Run with --stacktrace to get the full stack trace of this deprecation warning.)

> Task :backend:dependencyUpdates
The method ComponentSelectionRules.all(Closure) with injection of ComponentMetadata and/or IvyModuleDescriptor has been deprecated. This is scheduled to be removed in Gradle 6.0. Use the new methods on ComponentSelection instead.

> Task :backend:dependencyUpdates FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':backend:dependencyUpdates'.
> java.lang.StackOverflowError (no error message)

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```

The first deprecation warning is triggered by:
```
test {
    jacoco {
        append = true
        destinationFile = file("$buildDir/jacoco/jacocoTest.exec")
    }
}
```

The second seems to be triggered by gradle-versions-plugin itself, cause even if I remove:
```
dependencyUpdates {
    setResolutionStrategy {
        componentSelection {
            rules ->
                rules.all { ComponentSelection selection ->
                    boolean rejected = ['alpha', 'beta', 'rc', 'cr', 'm', 'preview']
                            .any { qualifier -> selection.candidate.version ==~ /(?i).*[.-]${qualifier}[.\d-]*/ }
                    if (rejected) {
                        selection.reject('Release candidate')
                    }
                }
        }
    }
}
```
the related warning still shows up.

**Comment by ben-manes** [comment]
The plugin's warning is fixed on master. In the initial bug, @danbar74 reproduced by using a dynamic version for the Liquibase dependency, which then failed to resolve in his IDE. This way we narrowed it down as not being due to the `dependencyUpdates` task and related to the their artifact's metadata or Gradle's resolution handing. If @danbar74 could attach their artifact's metadata, then it could probably be reproduced to determine where the fix should go.

**Comment by danbar74** [comment]
maven-metadata (version specific):
```xml
<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <groupId>org.liquibase</groupId>
  <artifactId>liquibase-core</artifactId>
  <version>3.7.0-SNAPSHOT</version>
  <versioning>
    <snapshot>
      <timestamp>20181015.043819</timestamp>
      <buildNumber>1</buildNumber>
    </snapshot>
    <lastUpdated>20181015043819</lastUpdated>
    <snapshotVersions>
      <snapshotVersion>
        <extension>jar</extension>
        <value>3.7.0-20181015.043819-1</value>
        <updated>20181015043819</updated>
      </snapshotVersion>
      <snapshotVersion>
        <extension>pom</extension>
        <value>3.7.0-20181015.043819-1</value>
        <updated>20181015043819</updated>
      </snapshotVersion>
    </snapshotVersions>
  </versioning>
</metadata>
```

maven-metadata (artifact)
```xml
<?xml version="1.0" encoding="UTF-8"?>
<metadata>
  <groupId>org.liquibase</groupId>
  <artifactId>liquibase-core</artifactId>
  <version>3.7.0-20181015.043819-1</version>
  <versioning>
    <latest>3.7.0-SNAPSHOT</latest>
    <versions>
      <version>3.7.0-SNAPSHOT</version>
    </versions>
    <lastUpdated>20181015043819</lastUpdated>
  </versioning>
</metadata>
```

[liquibase-core-3.7.0-20181015.043819-1.pom.zip](https://github.com/gradle/gradle/files/3087553/liquibase-core-3.7.0-20181015.043819-1.pom.zip)

**Comment by nkiesel** [comment]
is anyone working on a solution for this?  I just happily switched to Gradle 5.4.1 and dependencyUdates/buildSrcVersions does not work anymore.  I could provide more info like stacktraces and also try proposed solutions.

**Comment by nkiesel** [comment]
This is the stacktrace I get:
[stacktrace.log](https://github.com/gradle/gradle/files/3164309/stacktrace.log)

**Comment by ben-manes** [comment]
It should fail with a dynamic version `a:b:+` which is where `dependencyUpdates` comes into play an @danbar74 reproduced when used directly. However, this is due to a private dependency so it isn't a reproducible test case for the Gradle team. If you are using only public dependencies and can narrow it down to an small example, then I bet someone at Gradle would take a look.

**Comment by ljacomet** [comment]
Closing as this should have been fixed with the fix for #1084
